File: validator/engine/keyframes-parse-css.js
/**
 * @license DEDUPE_ON_MINIFY
 * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the license.
 */

goog.provide('parse_css.validateKeyframesCss');

goog.require('amp.validator.LIGHT');
goog.require('amp.validator.ValidationError');
goog.require('parse_css.ErrorToken');
goog.require('parse_css.RuleVisitor');
goog.require('parse_css.Stylesheet');
goog.require('parse_css.TRIVIAL_ERROR_TOKEN');


/**
 * Fills an ErrorToken with the provided position, code, and params.
 * @param {!parse_css.Token} positionToken
 * @param {!amp.validator.ValidationError.Code} code
 * @param {!Array<string>} params
 * @return {!parse_css.ErrorToken}
 */
function createErrorTokenAt(positionToken, code, params) {
  const token = new parse_css.ErrorToken(code, params);
  positionToken.copyPosTo(token);
  return token;
}

/** @private */
class KeyframesVisitor extends parse_css.RuleVisitor {
  /**
   * @param {!Array<parse_css.ErrorToken>} errors
   */
  constructor(errors) {
    super();

    /** @type {!Array<parse_css.ErrorToken>} */
    this.errors = errors;

    /** @type {boolean} */
    this.parentIsKeyframesAtRule = false;
  }

  /** @inheritDoc */
  visitQualifiedRule(qualifiedRule) {
    if (!this.parentIsKeyframesAtRule) {
      if (amp.validator.LIGHT) {
        this.errors.push(parse_css.TRIVIAL_ERROR_TOKEN);
      } else {
        this.errors.push(createErrorTokenAt(
            qualifiedRule,
            amp.validator.ValidationError.Code
                .CSS_SYNTAX_DISALLOWED_QUALIFIED_RULE_MUST_BE_INSIDE_KEYFRAME,
            ['style', qualifiedRule.ruleName()]));
      }
      return;
    }
    if (qualifiedRule.declarations.length > 0) return;
    if (amp.validator.LIGHT) {
      this.errors.push(parse_css.TRIVIAL_ERROR_TOKEN);
    } else {
      this.errors.push(createErrorTokenAt(
          qualifiedRule,
          amp.validator.ValidationError.Code
              .CSS_SYNTAX_QUALIFIED_RULE_HAS_NO_DECLARATIONS,
          ['style', qualifiedRule.ruleName()]));
    }
  }

  /** @inheritDoc */
  visitAtRule(atRule) {
    switch (atRule.name) {
      case 'keyframes':
      case '-moz-keyframes':
      case '-o-keyframes':
      case '-webkit-keyframes':
        if (this.parentIsKeyframesAtRule) {
          if (amp.validator.LIGHT) {
            this.errors.push(parse_css.TRIVIAL_ERROR_TOKEN);
          } else {
            this.errors.push(createErrorTokenAt(
                atRule,
                amp.validator.ValidationError.Code
                    .CSS_SYNTAX_DISALLOWED_KEYFRAME_INSIDE_KEYFRAME,
                ['style']));
          }
        }
        this.parentIsKeyframesAtRule = true;
        return;
      default:
    }
  }

  /** @inheritDoc */
  leaveAtRule(atRule) {
    this.parentIsKeyframesAtRule = false;
  }
}

/**
 * @param {!parse_css.Stylesheet} styleSheet
 * @param {!Array<!parse_css.ErrorToken>} errors
 */
parse_css.validateKeyframesCss = function(styleSheet, errors) {
  const visitor = new KeyframesVisitor(errors);
  styleSheet.accept(visitor);
};

File: validator/cpp/engine/keyframes-parse-css.cc
#include "cpp/engine/keyframes-parse-css.h"

#include "absl/memory/memory.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"

using absl::make_unique;
using amp::validator::ValidationError;
using htmlparser::css::AtRule;
using htmlparser::css::ErrorToken;
using htmlparser::css::QualifiedRule;
using htmlparser::css::RuleVisitor;
using htmlparser::css::Token;
using std::unique_ptr;
using std::vector;

namespace amp::validator::parse_css {
namespace {
// Fills an ErrorToken with the provided position, code, and params.
unique_ptr<ErrorToken> CreateParseErrorTokenAt(
    const Token& position_token, ValidationError::Code code,
    const std::vector<std::string>& params) {
  auto token = make_unique<ErrorToken>(code, params);
  position_token.CopyStartPositionTo(token.get());
  return token;
}

class KeyframesVisitor : public RuleVisitor {
 public:
  explicit KeyframesVisitor(std::vector<unique_ptr<ErrorToken>>* errors)
      : errors_(errors) {}

  // This is allowed:
  //   @keyframes anim1 { 100% {visibility: visible;} }
  // But these are not allowed:
  //   @media (min-width: 300px) { 100% {transform: translateX(-100%);} }
  //   amp-img { opacity: 0 }
  void VisitQualifiedRule(const QualifiedRule& qualified_rule) override {
    if (!parent_is_keyframes_at_rule_) {
      errors_->emplace_back(CreateParseErrorTokenAt(
          qualified_rule,
          ValidationError::
              CSS_SYNTAX_DISALLOWED_QUALIFIED_RULE_MUST_BE_INSIDE_KEYFRAME,
          {"style", qualified_rule.RuleName()}));
      return;
    }
    if (!qualified_rule.declarations().empty()) return;
    errors_->emplace_back(CreateParseErrorTokenAt(
        qualified_rule,
        ValidationError::CSS_SYNTAX_QUALIFIED_RULE_HAS_NO_DECLARATIONS,
        {"style", qualified_rule.RuleName()}));
  }

  void VisitAtRule(const AtRule& at_rule) override {
    // TODO: Move this to protoascii.
    if (at_rule.name() == "keyframes" || at_rule.name() == "-moz-keyframes" ||
        at_rule.name() == "-o-keyframes" ||
        at_rule.name() == "-webkit-keyframes") {
      if (parent_is_keyframes_at_rule_) {
        errors_->emplace_back(CreateParseErrorTokenAt(
            at_rule,
            ValidationError::CSS_SYNTAX_DISALLOWED_KEYFRAME_INSIDE_KEYFRAME,
            {"style"}));
      }
      parent_is_keyframes_at_rule_ = true;
    }
  }

  void LeaveAtRule(const AtRule& at_rule) override {
    parent_is_keyframes_at_rule_ = false;
  }

 private:
  std::vector<unique_ptr<ErrorToken>>* errors_;
  bool parent_is_keyframes_at_rule_ = false;
};
}  // namespace

void ValidateKeyframesCss(const htmlparser::css::Stylesheet& stylesheet,
                          std::vector<unique_ptr<ErrorToken>>* errors) {
  KeyframesVisitor visitor(errors);
  stylesheet.Accept(&visitor);
}
}  // namespace amp::validator::parse_css
